src/ring/adapter/jetty9.clj

```diff
@@ -44,39 +45,65 @@
     (string? response) {:body response}
     :else response))
 
+(defn- websocket-upgrade-response? [{:keys [status headers]}]
+  ;; HTTP 101 Switching Protocols
+  ;; https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/101
+  (and (= 101 status)
+       (let [headers (lower-case-keys headers)]
+         (and (= "websocket" (lower-case (get headers "upgrade")))
+              (= "upgrade" (lower-case (get headers "connection")))))))
+
+(defn ^:internal wrap-proxy-handler
+  "Wraps a Jetty handler in a ServletContextHandler.
+   
+   Websocket upgrades require a servlet context which makes it
+   necessary to wrap the handler in a servlet context handler."
+  [jetty-handler]
+  (doto (ServletContextHandler.)
+    (.setContextPath "/*")
+    (.setAllowNullPathInfo true)
+    (JettyWebSocketServletContainerInitializer/configure nil)
+    (.setHandler jetty-handler)))
+
 (defn ^:internal proxy-handler
   "Returns an Jetty Handler implementation for the given Ring handler."
   [handler]
-  (proxy [AbstractHandler] []
-    (handle [_ ^Request base-request ^HttpServletRequest request ^HttpServletResponse response]
-      (try
-        (let [request-map (build-request-map request)
-              response-map (-> (handler request-map)
-                               normalize-response)]
-          (when response-map
-            (servlet/update-servlet-response response response-map)))
-        (catch Throwable e
-          (.sendError response 500 (.getMessage e)))
-        (finally
-          (.setHandled base-request true))))))
+  (wrap-proxy-handler
+   (proxy [AbstractHandler] []
+     (handle [_ ^Request base-request ^HttpServletRequest request ^HttpServletResponse response]
+       (try
+         (let [request-map (build-request-map request)
+               response-map (-> (handler request-map)
+                                normalize-response)]
+           (when response-map
+             (if (websocket-upgrade-response? response-map)
```

Comment:
Use our new `ws-upgrade-request?` here?

Reply:
There's a difference between `ws-upgrade-request?` and `websocket-upgrade-response?`. The latter checks the request's http status code, whereas the former doesn't.